src/rust/engine/src/rule_graph.rs

```diff
@@ -1,14 +1,16 @@
 // Copyright 2017 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::collections::{hash_map, HashMap, HashSet};
+use std::collections::{hash_map, BTreeSet, HashMap, HashSet};
 use std::io;
 
 use core::{Function, Key, TypeConstraint, TypeId, Value, ANY_TYPE};
 use externs;
 use selectors::{Get, Select};
 use tasks::{Intrinsic, Task, Tasks};
 
+type ParamTypes = BTreeSet<TypeId>;
```

Comment:
What's the big difference between this and the `Params`?

Reply:
The key difference is `Params` contains `Key`s (ie, actual python values). But also, in theory performance is less critical in this case because the vast majority of entries should be created and destroyed during `RuleGraph` initialization.